documentation/porting_guidelines.md

```diff
@@ -0,0 +1,400 @@
+example when it is hard for us to do this in a generic way since device
+manufacturers all tend to do this in their own unique way and they are not very
+keen on sharing their low level boot details and security implementation with
+the rest of the world. This is especially true on ARMv7-A. For ARMv8-A it looks
+bit better, since ARM in ARM Trusted Firmware have implemented and defined how a
+abstract the chain of trust (see [auth-framework.md]). We have successfully
+verified OP-TEE by using the authentication framework from ARM Trusted Firmware
+(see [optee_with_auth_framework.md] for the details).
+
+## 6. Hardware Crypto IP
+By default OP-TEE uses a software crypto library (currently LibTomCrypt) and you
+have the ability to enable Crypto Extensions that was introduced with ARMv8-A
+(if the device is capable of that). To this date, we haven't had any hardware in
+our possession where we have had access to the real crypto hardware. Some of the
+devices we have in our hands do have hardware crypto IP's, but due to NDA's etc
+it hasn't been possible to enable it. If you have a device capable of doing
+crypto operations on a dedicated crypto block and you prefer to use that in
+favor for the software implementation, then you will need to implement a new
+`crypto_ops` structure and write the low level driver that communicates with the
+device. Our [crypto.md] file describes how to add and implement a new `struct
+crypto_ops`. Since the communication with crypto blocks tends to be quite
+different depending on what kind of crypto block you have, we haven't written
+how that should be done. It might be that we do that in the future when get hold
+of a device where we can use the crypto block.
+
+## 7. Power Management / PSCI
+In section 2 when we talked about the file `main.c`, we added a couple of
+handlers related to power management, we are talking about the following lines:
+```
+	.cpu_on = cpu_on_handler,
+	.cpu_off = pm_do_nothing,
+	.cpu_suspend = pm_do_nothing,
+	.cpu_resume = pm_do_nothing,
+	.system_off = pm_do_nothing,
+	.system_reset = pm_do_nothing,
+```
+The only function that actually does something there is the `cpu_on` function,
+the rest of them are stubbed. The main reason for that is because we think that
+how to suspend and resume is a device dependant thing. The code in OP-TEE is
+prepared so that callbacks etc from ARM Trusted Firmware will be routed to
+OP-TEE, but since the function(s) are just stubbed we won't do anything and just
+return. In a real production device, you would probably want to save and restore
+CPU states, secure hardware IPs' registers and TZASC and other memory firewall
+related setting when these callbacks are being called.
+
+## 8. Memory firewalls / TZASC
+ARM have defined a system IP / SoC peripheral called TrustZone Adress Space
+Controller (TZASC, see [TZC-380] and [TZC-400]). TZASC can be used configured
+DDR memory into separate regions in the address space, where each region can
```

Comment:
... separate regions in the **physical** address space ...

I think it's (sorry "it is") better to specify that this is not about virtual mapping/attributes.